test/e2e/user_configured_cabundle_test.go

```diff
@@ -0,0 +1,692 @@
+		}
+	}()
+
+	// create a server CA bundle for sample server to verify router certificate.
+	serviceCA := &corev1.ConfigMap{}
+	if err := kclient.Get(ctx, openshiftServerCA, serviceCA); err != nil {
+		t.Fatalf("failed to get configmap %q: %v", openshiftServerCA, err)
+	}
+	serverCA := serviceCA.Data["service-ca.crt"]
+
+	serverCABundleConfigMap := buildConfigMap(serverCABundleConfigMapName,
+		map[string]string{
+			"ca-bundle.crt": fmt.Sprintf("%s\n%s\n%s", serverCA, encodeCert(svcCACert1), encodeCert(svcCACert2)),
+		})
+	if err := kclient.Create(ctx, serverCABundleConfigMap); err != nil {
+		t.Fatalf("failed to create configmap %q: %v", serverCABundleConfigMapName, err)
+	}
+	defer func() {
+		if err := kclient.Delete(ctx, serverCABundleConfigMap); err != nil {
+			t.Fatalf("failed to delete configmap %q: %v", serverCABundleConfigMapName, err)
+		}
+	}()
+
+	// create a TLS secret for server.
+	svcTLSSecret1 := buildTLSSecret(svcTLSSecretName1, []byte(encodeCert(svcCert1)), []byte(encodeKey(svcKey1)))
+	if err := kclient.Create(ctx, svcTLSSecret1); err != nil {
+		t.Fatalf("failed to create secret %q: %v", svcTLSSecretName1, err)
+	}
+	defer func() {
+		if err := kclient.Delete(ctx, svcTLSSecret1); err != nil {
+			t.Fatalf("failed to delete secret %q: %v", svcTLSSecretName1, err)
+		}
+	}()
+
+	// create a TLS secret for server.
+	svcTLSSecret2 := buildTLSSecret(svcTLSSecretName2, []byte(encodeCert(svcCert2)), []byte(encodeKey(svcKey2)))
+	if err := kclient.Create(ctx, svcTLSSecret2); err != nil {
+		t.Fatalf("failed to create secret %q: %v", svcTLSSecretName2, err)
+	}
+	defer func() {
+		if err := kclient.Delete(ctx, svcTLSSecret2); err != nil {
+			t.Fatalf("failed to delete secret %q: %v", svcTLSSecretName2, err)
+		}
+	}()
+
+	// create a test server for verifying defaultDestinationCA scenarios.
+	svcDeployment1 := buildServerDeployment(svcDeploymentName1, svcLabels1, svcPort1, svcTLSSecretName1.Name, serverCABundleConfigMapName.Name)
+	if err := kclient.Create(ctx, svcDeployment1); err != nil {
+		t.Fatalf("failed to create deployment %q: %v", svcDeploymentName1, err)
+	}
+	defer func() {
+		if err := kclient.Delete(ctx, svcDeployment1); err != nil {
+			if errors.IsNotFound(err) {
+				return
+			}
+			t.Fatalf("failed to delete deployment %q: %v", svcDeploymentName1, err)
+		}
+	}()
+
+	// wait till deployment is available.
+	if err = wait.PollUntilContextTimeout(ctx, 2*time.Second, 3*time.Minute, true, func(ctx context.Context) (bool, error) {
+		if err := kclient.Get(ctx, svcDeploymentName1, svcDeployment1); err != nil {
+			t.Logf("failed to get server deployment %q: %v", svcDeploymentName1, err)
+			return false, nil
+		}
+		for _, cond := range svcDeployment1.Status.Conditions {
+			if cond.Type == appsv1.DeploymentAvailable {
+				return cond.Status == corev1.ConditionTrue, nil
+			}
+		}
+		return false, nil
+	}); err != nil {
+		t.Fatalf("timed out waiting for deployment %q to become ready: %v", svcDeploymentName1, err)
+	}
```

Comment:
This should probably be a helper function, like `waitForDeploymentComplete`.

Reply:
suggestion incorporated.